### ISSUE TITLE
Please add README or INSTALL file

### ISSUE THREAD

**Issue by CyberShadow** [i3/i3#2907]
There is currently no documentation in the project tree root on how to build i3. Please add a file describing the correct / supported way to do so.

If the documentation is elsewhere, please add a README file linking or directing readers to said location.

A README (or README.md) or INSTALL file is an established convention among open-source projects; it would be good if i3 followed suit.

Additionally, a README.md (or README) file will be automatically rendered by GitHub [on the project's main GitHub page](https://github.com/i3/i3), so it would be a good place to link to the project website and online documentation there.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by Airblader** [comment]
We generally dont want to encourage building from source but instead going with distribution packages. All the documentation is on the website or can be found in the docs folder, both of which are established standards as well. I don't see what benefit a README would add over that.

**Comment by CyberShadow** [comment]
> We generally dont want to encourage building from source but instead going with distribution packages.

That is no reason to not have a README file, even if it only just says that.

Building from source is required for contributing and troubleshooting (regression testing or debugging packaging issues), so it's also no reason to not document the correct procedure to build from source code.

> All the documentation is on the website or can be found in the docs folder,

The docs folder contains many files, including non-documentation files. It is not obvious which file describes the building process, if any.

> both of which are established standards as well. 

Unlike an INSTALL file, there is nothing on the website or the docs folder which can be immediately discovered when looking for build/install instructions.

> I don't see what benefit a README would add over that.

This issue is not a suggestion to add a README file as substitute for any existing documentation, but as an addition to it. Even a short README linking to places where existing documentation can be found would be preferred over nothing.

Either way, I have still not found any canonical documentation describing how to correctly build i3 from source; all that I've found is a mention in passing in `PACKAGE-MAINTAINER` and the `.travis.yml` file.

**Comment by Airblader** [comment]
We should address the request for a README and the issue of build instructions separately. For the README, I think it'd make sense to discuss an actual proposal over a PR. My only immediate input on it would be that we want to avoid duplication of documentation.

I think documenting the build instructions makes sense to some degree, but it should also be made clear who they are for and in which cases. Also they should live on the website as this is where the documentation lives. A separate issue would probably be appropriate for that (if it doesn't exist yet — I can't check right now).

@stapelberg What do you think?

**Comment by orestisf1993** [comment]
While we are at it I also propose to:

1. Symlink `CONTRIBUTING.md` to the project link because `.github` is hard to notice.
2. Mention the value of running xephyr for debugging and running code that is still being developed.

I can move these to a new issue.

**Comment by CyberShadow** [comment]
> For the README, I think it'd make sense to discuss an actual proposal over a PR.

I'd be glad to submit a PR with a proposed README file.

> Also they should live on the website as this is where the documentation lives. 

Documentation on how to build the source code may be better distributed with the source code itself, but I don't object. I understand the website is built from the `docs/` directory anyway.

> Symlink `CONTRIBUTING.md` to the project link because `.github` is hard to notice.

Why not just move the file to the project root instead? GitHub will pick it up from there as well, and that would make one less directory entry in total.

**Comment by CyberShadow** [comment]
Speaking of the `.github` directory, I'm not sure why `GOVERNANCE.md` is there, as it's not a file that GitHub recognizes. Seems like it was placed there in lieu of an existing place to put documentation targeted at contributors.

**Comment by stapelberg** [comment]
The documentation one should read before starting to hack on i3 is the hacking-howto, prominently linked in the “for developers” section on https://i3wm.org/docs/: https://i3wm.org/docs/hacking-howto.html

The (less discoverable) link for the git version of that document is http://build.i3wm.org/docs/hacking-howto.html

The hacking-howto already describes how to build i3 (https://i3wm.org/docs/hacking-howto.html#_how_to_build) and the testsuite documentation mentions Xephyr (https://i3wm.org/docs/testsuite.html#_installing_the_dependencies). Pull requests which point out that some people like to use Xephyr interactively are welcome.

Regarding a README file, I think it should just contain a link to https://i3wm.org/docs/, and possibly a link to .github/CONTRIBUTING.md. We never added one because we thought every user who considers contributing would be aware of the docs section on our website, but if such a hint helps contributors, I’m not against it.

Regarding the .github directory, I’d like to keep that as-is to avoid cluttering the root of our repository. It already contains too many files.

**Comment by CyberShadow** [comment]
> The hacking-howto already describes how to build i3

Wow. I did glance through the hacking guide, but it didn't occur to me to scroll all the way to the bottom to find the build instructions :) IMHO, as building the software is usually the first step to contributing (right after obtaining the source code), it may be worth reordering or otherwise making that section more visible.

Even when the build steps appear to be obvious (use the usual command for whatever build system the project seems to be using), knowing that the usual way is the correct way is helpful, as there may be unusual defaults that could cause problems later. (Case in point, it took me a while to discover `--disable-sanitizers` while bisecting i3 for #2909).

> Regarding a README file, I think it should just contain a link to https://i3wm.org/docs/, and possibly a link to .github/CONTRIBUTING.md.

-> https://github.com/i3/i3/pull/2910

> We never added one because we thought every user who considers contributing would be aware of the docs section on our website, but if such a hint helps contributors, I’m not against it.

Subjectively, it also looks kind of nice: https://github.com/CyberShadow/i3/tree/pull-20170909-033500#readme

**Comment by stapelberg** [comment]
> Wow. I did glance through the hacking guide, but it didn't occur to me to scroll all the way to the bottom to find the build instructions :) IMHO, as building the software is usually the first step to contributing (right after obtaining the source code), it may be worth reordering or otherwise making that section more visible.

Different people read differently. Personally, I would the hacking-howto from start to finish, discarding individual chapters as I determine they’re not relevant right now.

As you seem to have a clear spot in mind where a build instruction should have been, could you send a PR to move it there please? :)

> Even when the build steps appear to be obvious (use the usual command for whatever build system the project seems to be using), knowing that the usual way is the correct way is helpful, as there may be unusual defaults that could cause problems later. (Case in point, it took me a while to discover --disable-sanitizers while bisecting i3 for #2909).

I’m assuming you needed to disable sanitizers because the latest build environment changes introduce a new AddressSanitizer warning? I just noticed this, and sent a fix for it in PR #2911.

AddressSanitizers should be implementation details in the usual case and as such they are not worth mentioning explicitly, I think. I can see your point when you say they are unusual defaults, but I really wish they weren’t :).

**Comment by CyberShadow** [comment]
> Personally, I would the hacking-howto from start to finish, discarding individual chapters as I determine they’re not relevant right now.

Aah. I'm too impatient for that :) But bad habits aside, I think it would be nice to improve visibility of more important information. It's one thing when you have some spare time to contribute to a project and can leisurely peruse its documentation; another is when you're debugging why things broke after an update and just want to get them working again (without resorting to temporary kludges like downgrading software).

> As you seem to have a clear spot in mind where a build instruction should have been, could you send a PR to move it there please? :)

Will do!

> AddressSanitizers should be implementation details in the usual case and as such they are not worth mentioning explicitly, I think. I can see your point when you say they are unusual defaults, but I really wish they weren’t :).

In my case, I was getting this message after building i3 at the first bisect step between 4.13 and 4.14:

```
==12345==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
```

I followed the second suggestion in the error message, and ran i3 with `LD_PRELOAD=/usr/lib/libasan.so`. Though that got i3 to run, it caused subtle problems elsewhere: because LD_PRELOAD is inherited by child processes, it propagated to terminals started via i3 hotkeys, and commands launched from them. Not only did this cause ASan diagnostics from all sorts of software to be printed to stderr, but it caused some things to break due to ASan changing programs' exit code to 1 when it detected issues.

I'm sure there exist proper solutions to the above, but at that moment, I just wanted to continue the git bisection to get a clue about why things broke in 4.14 :)

**Comment by Airblader** [comment]
For the record, I also think build instructions belong to the top of such a document.

**Comment by stapelberg** [comment]
> ASan runtime does not come first in initial library list

We should look into fixing this, but I didn’t encounter the message on my machine. Can you share which asan version you use, and which linux kernel?

**Comment by CyberShadow** [comment]
> We should look into fixing this, but I didn’t encounter the message on my machine.

It doesn't happen any more with next or 4.14 (or 4.13 for that matter); I only encountered it with interim versions between 4.13 and 4.14 during git bisect. 

> Can you share which asan version you use, and which linux kernel?

```
$ pacman -Qo /usr/lib/libasan.so
/usr/lib/libasan.so is owned by gcc-libs-multilib 7.1.1-4

$ uname -a 
Linux home.thecybershadow.net 4.12.10-1-ARCH #1 SMP PREEMPT Wed Aug 30 12:18:42 CEST 2017 x86_64 GNU/Linux
```

**Comment by CyberShadow** [comment]
> As you seem to have a clear spot in mind where a build instruction should have been, could you send a PR to move it there please? :)

Done: https://github.com/i3/i3/pull/2928
